src/engine/utils/sort-entries-by-reference.ts

```diff
@@ -0,0 +1,115 @@
+import { some, filter, map, flatten, values, get, has } from 'lodash'
+import { AddedChangesetItem } from '../types'
+import { EntryField } from 'contentful'
+import { EntrySkeletonType } from 'contentful/dist/types/types/query'
+
+type LinksPerEntry = { index: number; linkIndexes: number[] }
+
+/**
+ * This function was inspired by the sortEntries function in contentful-import
+ * https://github.com/contentful/contentful-import/blob/028ba2cf1a0df9415da5f0608adb8a24e0428a98/lib/utils/sort-entries.ts
+ * Given a list of entries, this function reorders them so that entries which
+ * are linked from other entries always come first in the order. This ensures
+ * that when we publish the newly added entries, we are not publishing entries
+ * which contain links to other entries that haven't been published yet.
+ */
+export default function sortEntriesByReference(entries: AddedChangesetItem[]): AddedChangesetItem[] {
+  const linkedEntries = getLinkedEntries(entries)
+
+  const mergedLinkedEntries = mergeSort(linkedEntries, (a: LinksPerEntry) => {
+    return hasLinkedIndexesInFront(a)
+  })
+
+  return map(mergedLinkedEntries, (linkInfo: LinksPerEntry) => entries[linkInfo.index])
+
+  function hasLinkedIndexesInFront(item: LinksPerEntry): number {
+    if (hasLinkedIndexes(item)) {
+      return some(item.linkIndexes, (index) => index > item.index) ? 1 : -1
+    }
+    return 0
+  }
+
+  function hasLinkedIndexes(item: LinksPerEntry): boolean {
+    return item.linkIndexes.length > 0
+  }
+}
+
+function getLinkedEntries(entries: AddedChangesetItem[]): LinksPerEntry[] {
+  return map(entries, function (entry: AddedChangesetItem) {
+    const entryIndex = entries.indexOf(entry)
+
+    const rawLinks = map(entry.data.fields, (field) => {
+      field = values(field)[0]
+      if (isEntryLink(field)) {
+        return getFieldEntriesIndex(field, entries)
+      } else if (isEntityArray(field) && isEntryLink(field[0])) {
+        return map(field, (item: EntryField<EntrySkeletonType>) => getFieldEntriesIndex(item, entries))
+      }
+    })
+
+    return {
+      index: entryIndex,
+      linkIndexes: filter(flatten(rawLinks), (index: number) => index >= 0) as number[],
```

Comment:
[Array.prototype.flat()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) should do the same trick as `flatten`. 
[Array.prototype.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) should do the same trick as `filter`.